### ISSUE TITLE
Q&A

### ISSUE THREAD

**Issue by MisakiCoca** [JinyuanLiu-CV/TarDAL#5]
## About Train Code ｜ 有关代码的问题

The vast majority of the code currently in the repository is only a transitional version, and we will soon release the full version with following functions:

- Multiple training strategy (fuse, detect, fuse + detect)
- Load checkpoint from url & PyTorch Hub
- Build-in saliency detect network for data preprocessing
- AMP support

当前仓库内代码版本为过渡版本，我们将尽快发布具有如下功能的完整版本：

- 多种可选的训练策略 （融合模式，检测模式，联合模式）
- 自动从网络下载我们提供的预训练参数 & 使用 PyTorch Hub 一行命令调用我们的网络用于融合
- 内置预训练掩膜计算网络，以用于数据预处理
- 混合精度计算支持

## About The Datasets ｜ 有关训练集的划分方式

Details about how the dataset is divided will be given later.

数据集的具体划分方式我们将于之后详细公布。

**Comment by xiyunqiao** [comment]
Hi，Good work of your stucture in fusion with ir and vi，may I ask youranother arxiv: "Unsupervised Misaligned Infrared and Visible Image Fusion via Cross-Modality Image Generation and Registration" will be shown in future with the code in github?

**Comment by MisakiCoca** [comment]
> Hi，Good work of your stucture in fusion with ir and vi，may I ask youranother arxiv: "Unsupervised Misaligned Infrared and Visible Image Fusion via Cross-Modality Image Generation and Registration" will be shown in future with the code in github?

Sorry, we don't know. For more information on the paper you mentioned, please contact "diwang1211@mail.dlut.edu.cn".

**Comment by NYCXI** [comment]
Hi, thank you very much for your work, I have few questions to ask:
1.when will the divid details of the dataset will be released?(I have seen that you said it will be released soon few months ago, but it hasn't release until now)
2.when will the complete code will be released?
3.I didn‘t find scenario annotation information(eg: day, night; rain, haze) in M3FD_Detection/Annotation/.  Have I missed something?

**Comment by NYCXI** [comment]
> > Hi, thank you very much for your work, I have few questions to ask: 1.when will the divid details of the dataset will be released?(I have seen that you said it will be released soon few months ago, but it hasn't release until now) 2.when will the complete code will be released? 3.I didn‘t find scenario annotation information(eg: day, night; rain, haze) in M3FD_Detection/Annotation/. Have I missed something?
> 
> Hi, we deeply regret the poor service you received, but we are actively working to improve.
> 
> 1. At first, we didn't plan to divide the dataset in a strict way because we wanted to encourage researchers to use our data to do wonderful work without being limited by "SOTA". As the need has grown over time, one of our authors is now working hard to get this job done. @Rolling-Plain
> 2. In the new version of the code, we work on building a unified "fusion-detection" framework and try to set standards for future fusion dataset formats. So, it has a lot of work to do, and I will keep working on it after I finish the current work. Also, YOLOv5 is easy to use, so if you want to re-implement the results in the paper now, you can try yourself easily.
> 3. We have not yet disclosed the scenarios' classification information, so you have not missed anything. We are conducting experiments to divide training sets and testing sets in a manner that is more appropriate than simply by scene.
> 
> We appreciate your feedback and wish you the best of luck with your experiment. We will keep you apprised of any developments.

Thanks for your reply, and I get a few more questions here:

**Comment by NYCXI** [comment]
Thanks for your reply, I get one more questions here:
I try to run your training code, but it says I need a list.txt file, which seems to be a file about data organization? I would like to know what exactly is in it and how can I generate a file like this?
I checked the part of reading list.txt in fusion_data.py, but I don't quite understand the specific meaning of 'name', 'vsm', 's1' and 's2'.

**Comment by liufangcen** [comment]
感谢作者的贡献！
我在读代码时，没有找到联合训练的代码，并且论文中的相关描述我并不是特别明白。
请问联合训练模式的代码什么时候公布呀？

**Comment by hhuxuchang** [comment]
请问这个论文中的 saliency degree weight的求解代码 什么时候可以开源呢

**Comment by lemon256** [comment]
感谢作者的贡献，我发现无法进行自主的模型训练。我希望作者能给出data\train 文件夹中各文件的内容或者参考，即list.txt\mk\vsm 等几个文件，感谢

**Comment by rainySundayafternoon** [comment]
> > Hi, thank you very much for your work, I have few questions to ask: 1.when will the divid details of the dataset will be released?(I have seen that you said it will be released soon few months ago, but it hasn't release until now) 2.when will the complete code will be released? 3.I didn‘t find scenario annotation information(eg: day, night; rain, haze) in M3FD_Detection/Annotation/. Have I missed something?
> 
> Hi, we deeply regret the poor service you received, but we are actively working to improve.
> 
> 1. At first, we didn't plan to divide the dataset in a strict way because we wanted to encourage researchers to use our data to do wonderful work without being limited by "SOTA". As the need has grown over time, one of our authors is now working hard to get this job done. @Rolling-Plain
> 2. In the new version of the code, we work on building a unified "fusion-detection" framework and try to set standards for future fusion dataset formats. So, it has a lot of work to do, and I will keep working on it after I finish the current work. Also, YOLOv5 is easy to use, so if you want to re-implement the results in the paper now, you can try yourself easily.
> 3. We have not yet disclosed the scenarios' classification information, so you have not missed anything. We are conducting experiments to divide training sets and testing sets in a manner that is more appropriate than simply by scene.
> 
> We appreciate your feedback and wish you the best of luck with your experiment. We will keep you apprised of any developments.

I am very impressed by and appreciate the author's great work. But pardon my ignorance, I don't see why releasing the dataset split needs hard work and conducting further experiments for a new but different split. You have experimental results on each split, so why not just release the original split you used in your experiments?

**Comment by MisakiCoca** [comment]
@rainySundayafternoon 
Hello, and many thanks for your interest in our work!
Response to the query regarding the partitioning of datasets: The dataset utilized in the paper consists of 2,000 image pairs. We are still expanding and optimizing the dataset, including but not limited to adding, deleting, and optimizing the annotation, after the study has been submitted. During this procedure, we've introduced new categories and scenes, thus the numbers and labels of the old dataset used in the experiment no longer correlate to the most recent version released in March, and the old dataset is of limited value. If you have examined our dataset closely, you will see that our annotation is extremely accurate in comparison to other datasets in the field (e.g. Multispectral).
If you are a researcher in image fusion, the code we have published is adequate for comparative experimentation in this field. If you are seeking industrial use, as you describe, "easy-to-use YOLOv5 is already publically available", I believe you should be able to independently implement yourself version based the descriptions in our manuscript. In addition, this research falls under Low-level Vision (Multi-modality Image Processing), therefore it is unclear why you are comparing it to YOLOv5.
If you have any further inquiries, please feel free to contact us [mailto:rollingplainko@gmail.com] or [WeChat:RollingPlain].

**Comment by MisakiCoca** [comment]
@rainySundayafternoon
Hello, and many thanks for your interest in our work!

Response to the query regarding the partitioning of datasets: The dataset utilized in the paper consists of 2,000 image pairs. We are still expanding and optimizing the dataset, including but not limited to adding, deleting, and optimizing the annotation, after the study has been submitted. During this procedure, we've introduced new categories and scenes, thus the numbers and labels of the old dataset used in the experiment no longer correlate to the most recent version released in March, and the old dataset is of limited value. If you have examined our dataset closely, you will see that our annotation is extremely accurate in comparison to other datasets in the field (e.g. Multispectral).

If you are a researcher in image fusion, the code we have published is adequate for comparative experimentation in this field. If you are seeking industrial use, as you describe, "easy-to-use YOLOv5 is already publically available", I believe you should be able to independently implement yourself version based the descriptions in our manuscript. In addition, this research falls under Low-level Vision (Multi-modality Image Processing), therefore it is unclear why you are comparing it to YOLOv5.

If you have any further inquiries, please feel free to contact us [mailto:rollingplainko@gmail.com].

**Comment by rainySundayafternoon** [comment]
@MisakiCoca 

Thanks for spending time clarifying those.

I haven’t run the code myself and assumed it cannot be run based on the question above “when will the complete code will be released”. If that was my misunderstanding, I apologize. Also, regarding YOLOv5, I was just quoting what you said in your answers.

Thanks again.

**Comment by MisakiCoca** [comment]
@rainySundayafternoon 
We are pleased to have resolved your issue.
You can configure `Notifications` for this repository on GitHub to receive important updates on the work's progress in the future. You may also utilize our online notebook to obtain a preview of our image fusion. [CoLab](https://colab.research.google.com/github/JinyuanLiu-CV/TarDAL/blob/main/tutorial.ipynb)

**Comment by hit-1190200416** [comment]
你好，关于训练背景判别器的部分源代码里有这么一段
        real_s = self.gradient(self.gradient(vi) * mk)
        self.generator.eval()
        fake_s = self.gradient(self.generator(ir, vi) * (1 - mk))
请问这个地方对real_s的背景提取是不是应该改成(1 - mk)

**Comment by MisakiCoca** [comment]
@hit-1190200416 感谢您的指出！开源整理过程中错误，现已修正。
https://github.com/JinyuanLiu-CV/TarDAL/blob/7afdeb7ec6b94b67fb2a4f37db8fb5b3a07657b6/pipeline/train.py#L110-L113

**Comment by chaiquanshaonian** [comment]
训练文件不给的嘛

**Comment by JalehFar** [comment]
I tried to train the fusion model, but I got negative loss for the discriminators. I want to ask is it normal? As the Wasserstein divergence has been used and it is based on distance, how is it possible that it is negative?!

**Comment by misakicoca** [comment]
> I tried to train the fusion model, but I got negative loss for the discriminators. I want to ask is it normal? As the Wasserstein divergence has been used and it is based on distance, how is it possible that it is negative?!

Although it goes against intuition, the WGAN loss can indeed be negative, which is normal.

**Comment by fuhx-ai** [comment]
Thankd for your sharing! I found a small bug in the function (load_ckpt) in fusion.py when I tried to train the fusion model. The last two lines will not be executed because of the first line, so it cannot load discriminator parameters.

`def load_ckpt(self, ckpt: dict):
          ckpt = ckpt if 'fuse' not in ckpt else ckpt['fuse']
  
          # check eval mode
          if self.config.inference.use_eval is None:
              if 'use_eval' in ckpt:
                  logging.warning(f'overwriting inference.use_eval {self.config.inference.use_eval} with {ckpt["use_eval"]}')
                  self.config.inference.use_eval = ckpt['use_eval']
              else:
                  logging.warning(f'no use_eval settings found, using default (true)')
                  self.config.inference.use_eval = True
          if 'use_eval' in ckpt:
              ckpt.pop('use_eval')
  
          # load state dict
          self.generator.load_state_dict(ckpt)
          if self.mode == 'train' and 'disc' in ckpt:
              self.dis_t.load_state_dict(ckpt['disc']['t'])
              self.dis_d.load_state_dict(ckpt['disc']['d'])`

And the released models only include the generator's params so I didnot find the two discriminators'. Where can I find the discriminators' pre-trained models?

**Comment by fuhx-ai** [comment]
谢谢你的分享。我发现pipeline/fuse.py（81-99）中，由于第一行，最后两行将不会被执行。当执行else分支时，ckpt会改变，'disc' in ckpt 将一直是false，无法导入判别器的预训练参数。请问这是一个bug吗？
`ckpt = ckpt if 'fuse' not in ckpt else ckpt['fuse']`
官网提供的模型参数只包括生成器参数，不包括判别器的参数，请问我可以在哪里找到预训练的判别器呢？

**Comment by wyt19990913** [comment]
> 谢谢你的分享。我发现pipeline/fuse.py（81-99）中，由于第一行，最后两行将不会被执行。当执行else分支时，ckpt会改变，'disc' in ckpt 将一直是false，无法导入判别器的预训练参数。请问这是一个bug吗？ `ckpt = ckpt if 'fuse' not in ckpt else ckpt['fuse']` 官网提供的模型参数只包括生成器参数，不包括判别器的参数，请问我可以在哪里找到预训练的判别器呢？

请问您运行出结果了吗

**Comment by misakicoca** [comment]
> 谢谢你的分享。我发现pipeline/fuse.py（81-99）中，由于第一行，最后两行将不会被执行。当执行else分支时，ckpt会改变，'disc' in ckpt 将一直是false，无法导入判别器的预训练参数。请问这是一个bug吗？ `ckpt = ckpt if 'fuse' not in ckpt else ckpt['fuse']` 官网提供的模型参数只包括生成器参数，不包括判别器的参数，请问我可以在哪里找到预训练的判别器呢？

由于本工作的核心目的为「生成有利于人眼观测及计算机感知的融合图像」，因此判别器的参数我们并未保留，仅保留了生成器参数以供复现论文内展示数据。

**Comment by fuhx-ai** [comment]
> > 谢谢你的分享。我发现pipeline/fuse.py（81-99）中，由于第一行，最后两行将不会被执行。当执行else分支时，ckpt会改变，'disc' in ckpt 将一直是false，无法导入判别器的预训练参数。请问这是一个bug吗？ `ckpt = ckpt if 'fuse' not in ckpt else ckpt['fuse']` 官网提供的模型参数只包括生成器参数，不包括判别器的参数，请问我可以在哪里找到预训练的判别器呢？
> 
> 代码逻辑问题：感谢您的指出！这是一个编码错误，现已修改。
> 
> 判别器参数问题：由于本工作的核心目的为「生成有利于人眼观测及计算机感知的融合图像」，因此判别器的参数我们并未保留，仅保留了生成器参数以供复现论文内展示数据。

感谢您的耐心解答！

**Comment by fuhx-ai** [comment]
> > 谢谢你的分享。我发现pipeline/fuse.py（81-99）中，由于第一行，最后两行将不会被执行。当执行else分支时，ckpt会改变，'disc' in ckpt 将一直是false，无法导入判别器的预训练参数。请问这是一个bug吗？ `ckpt = ckpt if 'fuse' not in ckpt else ckpt['fuse']` 官网提供的模型参数只包括生成器参数，不包括判别器的参数，请问我可以在哪里找到预训练的判别器呢？
> 
> 请问您运行出结果了吗

我试着运行了这个程序，但是我的融合后检测的ap只有0.2~0.3，我不知道训练的哪个阶段出了问题
